File: src/unity/Runtime/Ads/Internal/Retrier.cs
using System;
using System.Threading.Tasks;

using Random = UnityEngine.Random;

namespace EE.Internal {
    using Processor = Func<Task<bool>>;

    internal class Retrier {
        private readonly float _baseDelay;
        private readonly float _multiplier;
        private readonly float _limit;
        private bool _processing;
        private bool _stopRequested;

        public Retrier(float baseDelay, float multiplier, float limit) {
            _baseDelay = baseDelay;
            _multiplier = multiplier;
            _limit = limit;
        }

        public async Task Process(Processor processor) {
            if (_processing) {
                return;
            }
            _processing = true;
            var delay = _baseDelay;
            while (true) {
                var x = Random.Range(0, 1);
                await Task.Delay((int) ((delay + x) * 1000));
                if (_stopRequested) {
                    _processing = false;
                    _stopRequested = false;
                    return;
                }
                var result = await processor();
                if (result) {
                    _processing = false;
                    return;
                }
                delay *= _multiplier;
                delay = Math.Min(delay, _limit);
            }
        }

        public void Stop() {
            if (_processing) {
                _stopRequested = true;
            }
        }
    }
}

File: src/cpp/ee/ads/internal/Retrier.cpp
#include "ee/ads/internal/Retrier.hpp"

#include <ee/core/Delay.hpp>
#include <ee/core/Random.hpp>
#include <ee/core/Task.hpp>

namespace ee {
namespace ads {
using Self = Retrier;

Self::Retrier(float base, float multiplier, float limit)
    : base_(base)
    , multiplier_(multiplier)
    , limit_(limit)
    , processing_(false)
    , stopRequested_(false) {}

Task<> Self::process(const Processor& processor) {
    if (processing_) {
        co_return;
    }
    processing_ = true;
    auto delay = base_;
    while (true) {
        auto x = random(0.0f, 1.0f);
        co_await Delay(delay + x);
        if (stopRequested_) {
            processing_ = false;
            stopRequested_ = false;
            co_return;
        }
        auto result = co_await processor();
        if (result) {
            processing_ = false;
            co_return;
        }
        delay *= multiplier_;
        delay = std::min(delay, limit_);
    }
}

void Self::stop() {
    if (processing_) {
        stopRequested_ = true;
    }
}
} // namespace ads
} // namespace ee
